velox/experimental/cudf/exec/ToCudf.cpp

```diff
@@ -0,0 +1,239 @@
+    return isAnyOf<exec::OrderBy>(op);
+  };
+
+  int32_t operatorsOffset = 0;
+  for (int32_t operatorIndex = 0; operatorIndex < operators.size();
+       ++operatorIndex) {
+    std::vector<std::unique_ptr<exec::Operator>> replaceOp;
+
+    exec::Operator* oper = operators[operatorIndex];
+    auto replacingOperatorIndex = operatorIndex + operatorsOffset;
+    VELOX_CHECK(oper);
+
+    const bool previousOperatorIsNotGpu =
+        (operatorIndex > 0 and !isSupportedGpuOperators[operatorIndex - 1]);
+    const bool nextOperatorIsNotGpu =
+        (operatorIndex < operators.size() - 1 and
+         !isSupportedGpuOperators[operatorIndex + 1]);
+
+    auto id = oper->operatorId();
+    if (previousOperatorIsNotGpu and acceptsGpuInput(oper)) {
+      auto planNode = getPlanNode(oper->planNodeId());
+      replaceOp.push_back(std::make_unique<CudfFromVelox>(
+          id, planNode->outputType(), ctx, planNode->id() + "-from-velox"));
+      replaceOp.back()->initialize();
+    }
+
+    if (auto* orderByOp = dynamic_cast<exec::OrderBy*>(oper)) {
+      auto id = orderByOp->operatorId();
+      auto planNode = std::dynamic_pointer_cast<const core::OrderByNode>(
+          getPlanNode(orderByOp->planNodeId()));
+      VELOX_CHECK(planNode != nullptr);
+      replaceOp.push_back(std::make_unique<CudfOrderBy>(id, ctx, planNode));
+      replaceOp.back()->initialize();
+    }
+
+    if (nextOperatorIsNotGpu and producesGpuOutput(oper)) {
+      auto planNode = getPlanNode(oper->planNodeId());
+      replaceOp.push_back(std::make_unique<CudfToVelox>(
+          id, planNode->outputType(), ctx, planNode->id() + "-to-velox"));
+      replaceOp.back()->initialize();
+    }
+
+    if (not replaceOp.empty()) {
+      operatorsOffset += replaceOp.size() - 1;
+      [[maybe_unused]] auto replaced = driverFactory_.replaceOperators(
+          driver_,
+          replacingOperatorIndex,
+          replacingOperatorIndex + 1,
+          std::move(replaceOp));
+      replacementsMade = true;
+    }
+  }
+
+  if (FLAGS_velox_cudf_debug) {
+    operators = driver_.operators();
+    std::cout << "Operators after adapting for cuDF: count ["
+              << operators.size() << "]" << std::endl;
+    for (auto& op : operators) {
+      std::cout << "  Operator: ID " << op->operatorId() << ": "
+                << op->toString() << std::endl;
+    }
+  }
+
+  return replacementsMade;
+}
+
+struct CudfDriverAdapter {
+  std::shared_ptr<rmm::mr::device_memory_resource> mr_;
+  std::shared_ptr<std::vector<core::PlanNodePtr>> planNodes_;
+
+  CudfDriverAdapter(std::shared_ptr<rmm::mr::device_memory_resource> mr)
+      : mr_(mr) {
+    planNodes_ = std::make_shared<std::vector<core::PlanNodePtr>>();
+  }
+
+  // Call operator needed by DriverAdapter
+  bool operator()(const exec::DriverFactory& factory, exec::Driver& driver) {
+    auto state = CompileState(factory, driver, *planNodes_);
+    // Stored planNodes_ from inspect.
+    auto res = state.compile();
```

Comment:
call `planNodes_->clear();` here can solve this issue https://github.com/facebookincubator/velox/pull/12735#issuecomment-2776646836

Reply:
I don't believe this is quite right. In plans involving multiple pipelines, the compile operator is called multiple times, once for each pipeline. So we need these stored plan nodes in all subsequent calls. Would adding a separate clear() method to the cudf driver adapter help? If you know when the task is finished, you can manually clear out the adapter.

Reply:
Do you mean the Task parallel execution mode? I read the wave DriverAdaper https://github.com/facebookincubator/velox/blob/main/velox/experimental/wave/exec/ToWave.cpp#L251, it does not need to store the plan nodes, is there any difference?

Reply:
Gluten uses the single thread task mode, so It's ok to clear it here. And if we have multiple drivers, after `createAndStartDrivers` in `Task::start`, can we clear the PlanNodes?

Is there any difference between the planNodes stored in DriverFactory and CompileState?
```
struct DriverFactory {
  std::vector<std::shared_ptr<const core::PlanNode>> planNodes;
```

Reply:
The adapters is saved to static zone, if we call the clear after the task is finished, we don't know which driver adapter is attached to the task. Gluten has parallel tasks in a single machine.

Reply:
>The adapters is saved to static zone, if we call the clear after the task is finished, we don't know which driver adapter is attached to the task. Gluten has parallel tasks in a single machine.

You wouldn't need to wait until a task are finished. You only need to wait until operator replacements have been made in it. 

To be clear, I am not standing my ground here. I just think we need a different solution to the one suggested.

Reply:
It looks good. I will verify this with all tpch queries. (with partition and hashjoin).

Reply:
I verified with tpch benchmarks. It works.